
### 📒 Step 1: Business Understanding

The aim of this project is to build a system that simplifies medical information retrieval for respiratory illnesses. Instead of relying on traditional machine learning models for prediction, the approach focuses on **data preprocessing and retrieval-based matching**.

The system will:

* Map **symptoms → disease → treatment** in a structured way.
* Convert the dataset into a clean, consistent CSV where each row represents a **unique disease–treatment pair with its associated symptoms**.
* Feed this structured data into a **vector database** (e.g., FAISS or Chroma) for efficient semantic search.
* Allow user queries (e.g., "shortness of breath and coughing") to be matched against the database, retrieving the most relevant disease and its treatment options.

**Why this approach?**

* Traditional predictive modeling is unnecessary here. Instead, the objective is **precise matching and retrieval** of knowledge.
* This method ensures **transparency and explainability**—users see exactly which symptoms map to which disease and treatment.
* It also supports cases where a disease has multiple treatments (each treatment is preserved as a separate entry).

This business framing highlights the project’s role as a **student-friendly, cost-free RAG (Retrieval-Augmented Generation) pipeline prototype** that prioritizes **usability, clarity, and reproducibility** over model complexity.



### 📒 Step 2: Data Understanding

In this step, we load the dataset into a Pandas DataFrame and perform an initial inspection. The goal is to understand:

* **Shape of the dataset** (number of rows and columns).
* **Column names** and what they represent.
* **Sample records** to see how symptoms, diseases, and treatments are structured.
* Whether the dataset contains missing values or irregularities that may affect preprocessing.

This understanding will guide the cleaning and preprocessing steps later.

In [1]:
# Import core libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Configure visualization style
sns.set(style="whitegrid")

# Load the dataset
file_path = "../data/respiratory symptoms and treatment.csv"  # adjust if needed
df = pd.read_csv(file_path)

# Display dataset structure
print("✅ Dataset loaded successfully\n")
print("Shape of dataset:", df.shape, "\n")
print("Columns:", df.columns.tolist(), "\n")
print("First 5 rows of the dataset:\n")
display(df.head())


✅ Dataset loaded successfully

Shape of dataset: (38537, 6) 

Columns: ['Symptoms', 'Age', 'Sex', 'Disease', 'Treatment', 'Nature'] 

First 5 rows of the dataset:



Symptoms  Age     Sex Disease    Treatment Nature
0                   coughing   5.0  female  Asthma   Omalizumab   high
1  tight feeling in the chest  4.0  female  Asthma  Mepolizumab   high
2                    wheezing  6.0    male  Asthma  Mepolizumab   high
3         shortness of breath  7.0    male  Asthma  Mepolizumab   high
4         shortness of breath  9.0    male  Asthma  Mepolizumab   high


### 📒 Step 2: Data Understanding (Results Explanation)

The dataset has been successfully loaded, containing **38,537 rows** and **6 columns**. Let’s break down what we now know:

* **Columns:**

  * **Symptoms** → text descriptions of medical complaints (e.g., “coughing”, “shortness of breath”).
  * **Age** → numeric (float), representing the age of the individual.
  * **Sex** → categorical (male/female).
  * **Disease** → categorical, the diagnosed condition (e.g., Asthma).
  * **Treatment** → categorical, prescribed medication or intervention (e.g., Omalizumab, Mepolizumab).
  * **Nature** → categorical (e.g., “high”), likely severity of the disease or treatment context.

* **Observations from sample rows:**

  * Multiple rows can belong to the same disease, but with different symptoms and treatments.
  * Example: *Asthma* appears multiple times, with symptoms like *“coughing”*, *“wheezing”*, and *“shortness of breath”*, and treatments like *Omalizumab* and *Mepolizumab*.
  * This confirms our earlier hypothesis: the dataset is **symptom-level granular**, not yet aggregated into the disease–treatment–symptom format we want.

* **Next Steps (Data Cleaning):**

  * Check for missing values and duplicates.
  * Normalize text entries (consistent casing, spacing).
  * Verify unique disease–treatment pairs.
  * Decide how to handle numeric fields like **Age** and categorical fields like **Sex** and **Nature** since the project’s retrieval pipeline may or may not need them.

This exploration shows the dataset is rich but fragmented—we’ll need to restructure it so that **all symptoms for a disease–treatment pair live in a single row** for embedding and retrieval.




### 📒 Step 3: Data Cleaning

**Markdown (Explanation before code):**

The purpose of this step is to ensure the dataset is consistent and reliable before we transform it. Data quality issues—such as missing values, duplicate entries, or inconsistent text formatting—can cause errors when we later aggregate symptoms or feed the data into a vector database.

Key cleaning checks include:

1. **Missing Values** → identify whether any column has null/NaN values.
2. **Duplicates** → check for repeated records that may artificially inflate results.
3. **Text Normalization** → standardize symptom, disease, treatment, and categorical columns (lowercase, strip spaces).
4. **Column Inspection** → decide whether all columns are needed for retrieval or only a subset (likely: `Disease`, `Treatment`, `Symptoms`).


In [2]:
# Check for missing values
print("🔎 Missing Values per Column:\n", df.isnull().sum(), "\n")

# Check for duplicate rows
print("🔎 Number of duplicate rows:", df.duplicated().sum(), "\n")

# Normalize text fields (Symptoms, Disease, Treatment, Sex, Nature)
text_columns = ["Symptoms", "Disease", "Treatment", "Sex", "Nature"]
for col in text_columns:
    df[col] = df[col].astype(str).str.lower().str.strip()

# Preview after normalization
print("✅ Preview after text normalization:\n")
display(df.head())


🔎 Missing Values per Column:
 Symptoms      696
Age           342
Sex           922
Disease       340
Treatment    2841
Nature       2190
dtype: int64 

🔎 Number of duplicate rows: 37634 

✅ Preview after text normalization:



Symptoms  Age     Sex Disease    Treatment Nature
0                    coughing  5.0  female  asthma   omalizumab   high
1  tight feeling in the chest  4.0  female  asthma  mepolizumab   high
2                    wheezing  6.0    male  asthma  mepolizumab   high
3         shortness of breath  7.0    male  asthma  mepolizumab   high
4         shortness of breath  9.0    male  asthma  mepolizumab   high

In [3]:
# Drop rows missing critical columns: Disease or Symptoms
df = df.dropna(subset=["Disease", "Symptoms"])

# Re-check missing values after dropping
missing_after = df.isna().sum()

print("🔎 Missing values after dropping rows without Disease or Symptoms:")
print(missing_after)
print(f"\n✅ Remaining rows: {len(df)}")


🔎 Missing values after dropping rows without Disease or Symptoms:
Symptoms       0
Age          342
Sex            0
Disease        0
Treatment      0
Nature         0
dtype: int64

✅ Remaining rows: 38537



* ✅ `Disease` and `Symptoms` are now **fully intact** (no missing values).
* ⚠️ Only `Age` still has **342 missing values** — this is acceptable for now, since `Age` is not the primary mapping key.
* ✅ `Sex`, `Treatment`, and `Nature` no longer report nulls (the NaNs were dropped along with their missing `Disease`/`Symptoms` rows).
* ✅ We retained **38,537 usable rows**, which is a strong base dataset.


In [5]:
# Assign the cleaned dataset to df_clean for consistency
df_clean = df.copy()

# Inspect unique values in categorical columns
categorical_cols = ['Sex', 'Disease', 'Treatment', 'Nature']

for col in categorical_cols:
    print(f"\n🔎 Unique values in {col}:")
    print(df_clean[col].value_counts(dropna=False))



🔎 Unique values in Sex:
Sex
male          21256
female        15411
not to say      948
nan             922
Name: count, dtype: int64

🔎 Unique values in Disease:
Disease
pneumonia                                6144
bronchitis                               4925
chronic obstructive pulmonary disease    3888
mesothelioma                             3216
pneumothorax                             2880
bronchiolitis                            2650
chronic bronchitis                       2016
bronchiectasis                           1950
influenza                                1872
tuberculosis                             1680
pulmonary hypertension                   1680
asthma                                   1096
chronic cough                             912
sleep apnea                               864
respiratory syncytial virus               720
acute respiratory distress syndrome       696
asbestosis                                504
aspergillosis                             504


Perfect — let’s lock down the **minor inconsistencies** in `Treatment` before we dive into the messy `Symptoms`.
We’ll build a **mapping dictionary** to unify values that are essentially the same but spelled differently.

In [6]:
# Define mapping dictionary for Treatment standardization
treatment_mapping = {
    'antibiotic': 'antibiotics',
    'antibiotics.': 'antibiotics',
    'oxyzen': 'oxygen',
    'consult doctor': 'consult a doctor',
    'inhealer': 'inhaler'
}

# Apply the mapping
df_clean['Treatment'] = df_clean['Treatment'].replace(treatment_mapping)

# Verify results after standardization
print("🔎 Unique treatments after normalization:")
print(df_clean['Treatment'].value_counts().head(20))


🔎 Unique treatments after normalization:
Treatment
antibiotics                          9839
chemotherapy                         2928
isotonic sodium chloride solution    2880
nan                                  2841
consult a doctor                     2336
oseltamivir                          1872
saline nose drops                    1800
oxygen                               1704
diuretics                            1680
pulmonary rehabilitation             1104
cough medicine                        960
inhaler                               957
hypertonic saline                     850
adaptive servo-ventilation            816
ethambutol                            720
intravenous fluids                    720
steroids to reduce inflammation       672
x-ray                                 624
pyrazinamide                          528
surgery                               432
Name: count, dtype: int64


**Treatment** column is standardized:

* `antibiotic` + `antibiotics.` merged into **antibiotics** (9839 total)
* `oxyzen` corrected to **oxygen** (1704 total)
* `consult doctor` unified with **consult a doctor** (2336 total)
* `inhealer` corrected to **inhaler** (957 total)

This means treatments are now clean and won’t fragment your later grouping or aggregation.

👉 Next step: **Symptoms cleanup**.
Unlike Treatments, Symptoms are free-text and will need:

1. **Lowercasing and stripping** (already done).
2. **Deduplication of similar terms** (e.g., “shortness of breath” vs “breathlessness”).
3. **Optionally lemmatization** (e.g., “coughing” → “cough”).
4. Grouping similar variations into a **controlled vocabulary** for consistency.


In [8]:
# Count unique symptoms
num_unique_symptoms = df_clean['Symptoms'].nunique()
print(f"🔎 Number of unique symptoms: {num_unique_symptoms}")

# Preview some of them
print("\n✅ Sample of unique symptoms:")
print(df_clean['Symptoms'].unique())  # first 50 unique symptoms


🔎 Number of unique symptoms: 78

✅ Sample of unique symptoms:
['coughing' 'tight feeling in the chest' 'wheezing' 'shortness of breath'
 'fever' 'cold' 'allergy' 'coughing up yellow or green mucus daily'
 'shortness of breath that gets worse during flare-ups'
 'fatigue, feeling run-down or tired' 'chest pain'
 'whistling sound while you breathe' 'coughing up blood' 'runny nose'
 'stuffy nose' 'loss of appetite' 'cough' 'low-grade fever'
 'chest congestion' 'whistling sound while breathing' 'yellow cough'
 'feeling run-down or tired' 'mucus' 'nan' 'chronic cough' 'fatigue'
 'lower back pain' 'dry cough' 'greenish cough' 'cough with blood'
 'sweating' 'shaking' 'rapid breathing' 'shallow breathing' 'low energy'
 'nausea' 'vomiting' 'sharp chest pain' 'bluish skin' 'rapid heartbeat'
 'high fever' 'headache' 'muscle aches' 'joint pain' 'chills'
 'sore throat' 'nasal congestion' 'diarrhea' 'breath' 'dizziness'
 'fainting' 'heart palpitations' 'edema' 'snoring' 'daytime sleepiness'
 'pauses 

In [9]:
# List all unique symptoms
unique_symptoms = df_clean['Symptoms'].unique()

print("🔎 List of all unique symptoms:\n")
for i, symptom in enumerate(unique_symptoms, 1):
    print(f"{i}. {symptom}")


🔎 List of all unique symptoms:

1. coughing
2. tight feeling in the chest
3. wheezing
4. shortness of breath
5. fever
6. cold
7. allergy
8. coughing up yellow or green mucus daily
9. shortness of breath that gets worse during flare-ups
10. fatigue, feeling run-down or tired
11. chest pain
12. whistling sound while you breathe
13. coughing up blood
14. runny nose
15. stuffy nose
16. loss of appetite
17. cough
18. low-grade fever
19. chest congestion
20. whistling sound while breathing
21. yellow cough
22. feeling run-down or tired
23. mucus
24. nan
25. chronic cough
26. fatigue
27. lower back pain
28. dry cough
29. greenish cough
30. cough with blood
31. sweating
32. shaking
33. rapid breathing
34. shallow breathing
35. low energy
36. nausea
37. vomiting
38. sharp chest pain
39. bluish skin
40. rapid heartbeat
41. high fever
42. headache
43. muscle aches
44. joint pain
45. chills
46. sore throat
47. nasal congestion
48. diarrhea
49. breath
50. dizziness
51. fainting
52. heart palpitatio

this symptom set is full of:

* **Duplicates / Variations**:

  * “cough” vs “coughing” vs “dry cough” vs “chronic cough” vs “cough with blood” vs “wheezing cough” vs “persistent dry coug” (typo).
  * “shortness of breath” vs “short of breath” vs “breath”.
  * “tight feeling in the chest” vs “chest tightness or chest pain” vs “chest pain”.
  * “whistling sound while you breathe” vs “whistling sound while breathing”.

* **Synonyms or near-duplicates**:

  * “fatigue” vs “feeling run-down or tired” vs “fatigue, feeling run-down or tired”.
  * “low energy” overlaps with “fatigue”.
  * “loss of appetite” vs “loss of appetite and unintentional weight loss” vs “weight loss from loss of appetite”.

* **Compound symptoms**:

  * “fatigue, feeling run-down or tired” should be split or reduced to “fatigue”.
  * “short, shallow and rapid breathing” overlaps with “rapid breathing” and “shallow breathing”.

* **Errors**:

  * “persistent dry coug” → typo, should be “persistent dry cough”.

Here’s the play:

1. Build a **mapping dictionary** to collapse all these variations into standardized forms.
2. Keep **atomic, distinct symptoms** (like “edema” or “snoring”) separate.
3. Remove “nan” entries (they’re just missing values).

In [10]:
df_clean.head(20)  # Display the first few rows of the cleaned DataFrame

Symptoms   Age     Sex Disease     Treatment  Nature
0                     coughing   5.0  female  asthma    omalizumab    high
1   tight feeling in the chest   4.0  female  asthma   mepolizumab    high
2                     wheezing   6.0    male  asthma   mepolizumab    high
3          shortness of breath   7.0    male  asthma   mepolizumab    high
4          shortness of breath   9.0    male  asthma   mepolizumab    high
5   tight feeling in the chest   NaN    male  asthma   mepolizumab    high
6          shortness of breath   NaN    male  asthma   mepolizumab    high
7   tight feeling in the chest   8.0  female  asthma   mepolizumab    high
8          shortness of breath  36.0  female  asthma   mepolizumab  medium
9                     wheezing  40.0  female  asthma    omalizumab  medium
10                    coughing  46.0  female  asthma    omalizumab  medium
11         shortness of breath  56.0  female  asthma    omalizumab  medium
12         shortness of breath  58.0  female  asthma    omalizumab  medium
13         shortness of breath  67.0  female  asthma    omalizumab    high
14                    coughing  87.0    male  asthma    omalizumab    high
15         shortness of breath  65.0    male  asthma   mepolizumab    high
16  tight feeling in the chest  54.0  female  asthma   mepolizumab  medium
17                    wheezing  34.0  female  asthma    omalizumab  medium
18         shortness of breath  78.0  female  asthma  itraconazole    high
19         shortness of breath   7.0  female  asthma       inhaler    high

In [13]:

# Normalize the raw symptom column
df_clean['symptom_clean'] = df_clean['Symptoms'].str.lower().str.strip()

# --- Step 1: Define a standardization dictionary ---
symptom_map = {
    # coughing & variants
    "coughing": "cough",
    "cough": "cough",
    "dry cough": "dry cough",
    "chronic cough": "chronic cough",
    "wheezing cough": "wheezing cough",
    "persistent dry coug": "persistent dry cough",  # spelling correction
    "persistent dry cough": "persistent dry cough",
    "a cough that lasts more than three weeks": "persistent cough",
    "yellow cough": "cough with yellow mucus",
    "greenish cough": "cough with green mucus",
    "cough with blood": "cough with blood",
    "coughing up blood": "cough with blood",
    "coughing up yellow or green mucus daily": "productive cough with yellow/green mucus",
    
    # breathing issues
    "shortness of breath": "shortness of breath",
    "short of breath": "shortness of breath",
    "breath": "shortness of breath",
    "shortness of breath that gets worse during flare-ups": "shortness of breath (worsens during flare-ups)",
    "rapid breathing": "rapid breathing",
    "shallow breathing": "shallow breathing",
    "short, shallow and rapid breathing": "rapid shallow breathing",
    "whistling sound while you breathe": "wheezing",
    "whistling sound while breathing": "wheezing",
    "wheezing": "wheezing",
    "a dry, crackling sound in the lungs while breathing in": "lung crackles",
    
    # chest-related
    "tight feeling in the chest": "chest tightness",
    "chest tightness or chest pain": "chest tightness/pain",
    "chest pain": "chest pain",
    "sharp chest pain": "sharp chest pain",
    "chest congestion": "chest congestion",
    
    # fever variants
    "fever": "fever",
    "low-grade fever": "low-grade fever",
    "high fever": "high fever",
    
    # fatigue & low energy
    "fatigue": "fatigue",
    "fatigue, feeling run-down or tired": "fatigue",
    "feeling run-down or tired": "fatigue",
    "low energy": "fatigue",
    
    # nose/throat
    "runny nose": "runny nose",
    "stuffy nose": "nasal congestion",
    "nasal congestion": "nasal congestion",
    "sore throat": "sore throat",
    
    # appetite/weight
    "loss of appetite": "loss of appetite",
    "loss of appetite and unintentional weight loss": "loss of appetite & weight loss",
    "weight loss": "weight loss",
    "weight loss from loss of appetite": "weight loss from appetite loss",
    
    # systemic / flu-like
    "cold": "cold",
    "allergy": "allergy",
    "muscle aches": "muscle aches",
    "joint pain": "joint pain",
    "headache": "headache",
    "morning headaches": "morning headaches",
    "chills": "chills",
    "sweating": "sweating",
    "shaking": "shaking",
    "night sweats": "night sweats",
    
    # GI issues
    "nausea": "nausea",
    "vomiting": "vomiting",
    "diarrhea": "diarrhea",
    
    # cardio / circulation
    "rapid heartbeat": "rapid heartbeat",
    "faster heart beating": "rapid heartbeat",
    "heart palpitations": "heart palpitations",
    "edema": "edema",
    "bluish skin": "cyanosis",
    "dizziness": "dizziness",
    "fainting": "fainting",
    
    # sleep & cognitive
    "snoring": "snoring",
    "daytime sleepiness": "daytime sleepiness",
    "pauses in breathing": "sleep apnea episodes",
    "frequently waking": "frequent waking",
    "dry mouth": "dry mouth",
    "difficulties with memory and concentration": "memory/concentration problems",
    "unusual moodiness": "mood changes",
    "irritability": "irritability",
    
    # rare
    "wider and rounder than normal fingertips and toes": "clubbing",
    "distressing": "distress",
    "pain": "pain",
    "nan": None  # treat missing
}

# --- Step 2: Apply mapping ---
df_clean['symptom_standardized'] = (
    df_clean['symptom_clean'].map(symptom_map).fillna(df_clean['symptom_clean'])
)

# --- Step 3: Check results ---
print(
    df_clean[['Symptoms', 'symptom_standardized']]
    .drop_duplicates()
    .sort_values(by='symptom_standardized')
)


                              Symptoms            symptom_standardized
35                             allergy                         allergy
231                   chest congestion                chest congestion
69                          chest pain                      chest pain
1           tight feeling in the chest                 chest tightness
980      chest tightness or chest pain            chest tightness/pain
..                                 ...                             ...
981  weight loss from loss of appetite  weight loss from appetite loss
77   whistling sound while you breathe                        wheezing
2                             wheezing                        wheezing
233    whistling sound while breathing                        wheezing
899                     wheezing cough                  wheezing cough

[78 rows x 2 columns]


In [14]:
df_clean.head(20)  # Display the first few rows of the cleaned DataFrame

Symptoms   Age     Sex Disease     Treatment  Nature  \
0                     coughing   5.0  female  asthma    omalizumab    high   
1   tight feeling in the chest   4.0  female  asthma   mepolizumab    high   
2                     wheezing   6.0    male  asthma   mepolizumab    high   
3          shortness of breath   7.0    male  asthma   mepolizumab    high   
4          shortness of breath   9.0    male  asthma   mepolizumab    high   
5   tight feeling in the chest   NaN    male  asthma   mepolizumab    high   
6          shortness of breath   NaN    male  asthma   mepolizumab    high   
7   tight feeling in the chest   8.0  female  asthma   mepolizumab    high   
8          shortness of breath  36.0  female  asthma   mepolizumab  medium   
9                     wheezing  40.0  female  asthma    omalizumab  medium   
10                    coughing  46.0  female  asthma    omalizumab  medium   
11         shortness of breath  56.0  female  asthma    omalizumab  medium   
12         shortness of breath  58.0  female  asthma    omalizumab  medium   
13         shortness of breath  67.0  female  asthma    omalizumab    high   
14                    coughing  87.0    male  asthma    omalizumab    high   
15         shortness of breath  65.0    male  asthma   mepolizumab    high   
16  tight feeling in the chest  54.0  female  asthma   mepolizumab  medium   
17                    wheezing  34.0  female  asthma    omalizumab  medium   
18         shortness of breath  78.0  female  asthma  itraconazole    high   
19         shortness of breath   7.0  female  asthma       inhaler    high   

                 symptom_clean symptom_standardized  
0                     coughing                cough  
1   tight feeling in the chest      chest tightness  
2                     wheezing             wheezing  
3          shortness of breath  shortness of breath  
4          shortness of breath  shortness of breath  
5   tight feeling in the chest      chest tightness  
6          shortness of breath  shortness of breath  
7   tight feeling in the chest      chest tightness  
8          shortness of breath  shortness of breath  
9                     wheezing             wheezing  
10                    coughing                cough  
11         shortness of breath  shortness of breath  
12         shortness of breath  shortness of breath  
13         shortness of breath  shortness of breath  
14                    coughing                cough  
15         shortness of breath  shortness of breath  
16  tight feeling in the chest      chest tightness  
17                    wheezing             wheezing  
18         shortness of breath  shortness of breath  
19         shortness of breath  shortness of breath

In [15]:
# Check total number of rows in df_clean
print(f"Number of rows in df_clean: {len(df_clean)}")


Number of rows in df_clean: 38537


In [17]:
# 🔎 Step 1: Validate Standardization

# Check standardized symptoms
print("🔎 Number of unique standardized symptoms:", df_clean['symptom_standardized'].nunique())
print("\nSample of standardized symptoms:")
print(df_clean['symptom_standardized'].dropna().unique())  # show first 30 unique values

# Check standardized treatments
print("\n🔎 Number of unique treatments:", df_clean['Treatment'].nunique())
print("\nUnique treatments:")
print(df_clean['Treatment'].unique())


🔎 Number of unique standardized symptoms: 66

Sample of standardized symptoms:
['cough' 'chest tightness' 'wheezing' 'shortness of breath' 'fever' 'cold'
 'allergy' 'productive cough with yellow/green mucus'
 'shortness of breath (worsens during flare-ups)' 'fatigue' 'chest pain'
 'cough with blood' 'runny nose' 'nasal congestion' 'loss of appetite'
 'low-grade fever' 'chest congestion' 'cough with yellow mucus' 'mucus'
 'nan' 'chronic cough' 'lower back pain' 'dry cough'
 'cough with green mucus' 'sweating' 'shaking' 'rapid breathing'
 'shallow breathing' 'nausea' 'vomiting' 'sharp chest pain' 'cyanosis'
 'rapid heartbeat' 'high fever' 'headache' 'muscle aches' 'joint pain'
 'chills' 'sore throat' 'diarrhea' 'dizziness' 'fainting'
 'heart palpitations' 'edema' 'snoring' 'daytime sleepiness'
 'sleep apnea episodes' 'memory/concentration problems' 'mood changes'
 'irritability' 'frequent waking' 'morning headaches' 'dry mouth'
 'wheezing cough' 'rapid shallow breathing' 'persistent coug


### 🔎 Interpretation of Standardization Results

1. **Symptoms**

   * We successfully reduced the noisy raw 78 unique symptoms down to **66 standardized symptoms**.
   * Variations such as spelling errors, synonyms, and near-duplicates were consolidated (e.g., *“coughing up blood”* and *“cough with blood”* now map to the same label).
   * Informative variations were **preserved** (e.g., *“dry cough”* vs *“chronic cough”* vs *“productive cough with yellow/green mucus”*).
   * Remaining `nan` entries represent missing or unrecorded symptoms.

2. **Treatments**

   * Treatment options were consolidated into **27 unique values**.
   * Inconsistencies such as *“antibiotic”*, *“antibiotics.”*, etc., were standardized into a single category **“antibiotics”**.
   * Duplicates like *“consult doctor”* vs *“consult a doctor”* were unified.
   * Some still remain descriptive (e.g., *“stay away from cold places”*) but that’s fine, as these represent actual treatment variations.

3. **Readiness for Preprocessing**

   * The dataset is now **clean enough to begin restructuring** into the **RAG-friendly format** (disease → all associated symptoms + treatment options).
   * Next, we’ll handle **missing values** (mainly in `Age`, `Nature`, and the `nan` in `symptom_standardized` and `Treatment`)  

In [18]:
df_clean.columns.tolist()

['Symptoms',
 'Age',
 'Sex',
 'Disease',
 'Treatment',
 'Nature',
 'symptom_clean',
 'symptom_standardized']

In [19]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38537 entries, 0 to 38536
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Symptoms              38537 non-null  object 
 1   Age                   38195 non-null  float64
 2   Sex                   38537 non-null  object 
 3   Disease               38537 non-null  object 
 4   Treatment             38537 non-null  object 
 5   Nature                38537 non-null  object 
 6   symptom_clean         38537 non-null  object 
 7   symptom_standardized  38537 non-null  object 
dtypes: float64(1), object(7)
memory usage: 2.4+ MB



### 🔧 Transformation Steps

1. **Drop irrelevant or missing entries** (`nan` treatments/symptoms won’t be useful in the vector DB).
2. **Group by (Disease, Treatment)**.

   * This ensures each treatment option for a disease has its own row.
3. **Aggregate all symptoms for each disease-treatment pair** into one text field.

   * Example: `"cough, fever, fatigue"`.
4. **Clean & standardize symptoms** (remove duplicates, strip spaces, lowercasing where necessary).
5. **Produce final DataFrame/CSV** with exactly **3 columns**:

   * `Disease`
   * `Symptoms` (comma-separated string)
   * `Treatment`

---

### ✅ Example Output (Restructured)

| Disease      | Symptoms                                                    | Treatment                          |
| ------------ | ----------------------------------------------------------- | ---------------------------------- |
| Asthma       | cough, chest tightness, wheezing, shortness of breath       | inhaler                            |
| Asthma       | cough, chest tightness, wheezing, shortness of breath       | omalizumab                         |
| Tuberculosis | persistent cough, night sweats, weight loss, chest pain     | rifampin, pyrazinamide, ethambutol |
| Influenza    | fever, chills, headache, fatigue, sore throat, muscle aches | oseltamivir                        |
| Pneumonia    | fever, cough with yellow mucus, chest pain, fatigue         | antibiotics                        |



In [20]:

# ✅ Start with only the relevant columns
df_ready = df_clean[['Age', 'Sex', 'symptom_standardized', 'Disease', 'Treatment']].dropna()

# Normalize text fields to keep consistency
df_ready['Disease'] = df_ready['Disease'].str.strip()
df_ready['Treatment'] = df_ready['Treatment'].str.strip().str.lower()
df_ready['symptom_standardized'] = df_ready['symptom_standardized'].str.strip().str.lower()

# Group by Disease + Treatment + Age + Sex, aggregate symptoms
df_grouped = (
    df_ready
    .groupby(['Disease', 'Treatment', 'Age', 'Sex'])['symptom_standardized']
    .apply(lambda x: ', '.join(sorted(set(x))))  # keep unique, sorted
    .reset_index()
)

# Rename for clarity
df_grouped = df_grouped.rename(columns={
    'symptom_standardized': 'Symptoms'
})

In [21]:
df_grouped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 761 entries, 0 to 760
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Disease    761 non-null    object 
 1   Treatment  761 non-null    object 
 2   Age        761 non-null    float64
 3   Sex        761 non-null    object 
 4   Symptoms   761 non-null    object 
dtypes: float64(1), object(4)
memory usage: 29.9+ KB


In [24]:
df_grouped.head(50)  # Display the first few rows of the final DataFrame

Disease                   Treatment   Age  \
0   acute respiratory distress syndrome            consult a doctor   3.0   
1   acute respiratory distress syndrome            consult a doctor   6.0   
2   acute respiratory distress syndrome            consult a doctor   9.0   
3   acute respiratory distress syndrome            consult a doctor  13.0   
4   acute respiratory distress syndrome            consult a doctor  18.0   
5   acute respiratory distress syndrome            consult a doctor  20.0   
6   acute respiratory distress syndrome            consult a doctor  21.0   
7   acute respiratory distress syndrome            consult a doctor  25.0   
8   acute respiratory distress syndrome            consult a doctor  27.0   
9   acute respiratory distress syndrome            consult a doctor  35.0   
10  acute respiratory distress syndrome            consult a doctor  39.0   
11  acute respiratory distress syndrome            consult a doctor  42.0   
12  acute respiratory distress syndrome            consult a doctor  43.0   
13  acute respiratory distress syndrome            consult a doctor  43.0   
14  acute respiratory distress syndrome            consult a doctor  50.0   
15  acute respiratory distress syndrome            consult a doctor  54.0   
16  acute respiratory distress syndrome            consult a doctor  55.0   
17  acute respiratory distress syndrome            consult a doctor  59.0   
18  acute respiratory distress syndrome            consult a doctor  60.0   
19  acute respiratory distress syndrome            consult a doctor  65.0   
20  acute respiratory distress syndrome            consult a doctor  67.0   
21  acute respiratory distress syndrome            consult a doctor  68.0   
22  acute respiratory distress syndrome            consult a doctor  70.0   
23  acute respiratory distress syndrome            consult a doctor  72.0   
24  acute respiratory distress syndrome            consult a doctor  78.0   
25  acute respiratory distress syndrome            consult a doctor  85.0   
26  acute respiratory distress syndrome            consult a doctor  86.0   
27  acute respiratory distress syndrome            consult a doctor  90.0   
28                           asbestosis            consult a doctor   1.0   
29                           asbestosis            consult a doctor  12.0   
30                           asbestosis            consult a doctor  13.0   
31                           asbestosis                      oxygen   2.0   
32                           asbestosis                      oxygen   4.0   
33                           asbestosis                      oxygen  66.0   
34                           asbestosis                      oxygen  76.0   
35                           asbestosis                      oxygen  78.0   
36                           asbestosis                      oxygen  89.0   
37                           asbestosis  stay away from cold places   8.0   
38                           asbestosis  stay away from cold places   9.0   
39                           asbestosis  stay away from cold places  11.0   
40                           asbestosis  stay away from cold places  15.0   
41                           asbestosis  stay away from cold places  20.0   
42                           asbestosis  stay away from cold places  25.0   
43                           asbestosis  stay away from cold places  42.0   
44                           asbestosis  stay away from cold places  47.0   
45                        aspergillosis                itraconazole   4.0   
46                        aspergillosis                itraconazole  11.0   
47                        aspergillosis                itraconazole  19.0   
48                        aspergillosis                itraconazole  21.0   
49                        aspergillosis                itraconazole  45.0   

           Sex                            Symptoms  
0          nan                     rapid heartbeat 

Including **Age** and **Sex** fragments the dataset unnecessarily because the symptoms aren’t exclusive to one group. If your end-goal is a **vectorized knowledge base**, then keeping them only adds noise.

So the fix is:

* Drop `Age` and `Sex` entirely.
* Group only by `Disease + Treatment`.
* Aggregate all unique symptoms into one row per **disease-treatment pair**.




In [28]:
# ✅ Start with only the key columns
df_ready = df_clean[['symptom_standardized', 'Disease', 'Treatment']].dropna()

# Normalize text fields
df_ready['Disease'] = df_ready['Disease'].str.strip()
df_ready['Treatment'] = df_ready['Treatment'].str.strip().str.lower()
df_ready['symptom_standardized'] = df_ready['symptom_standardized'].str.strip().str.lower()

# ✅ Group by Disease + Treatment only
df_grouped = (
    df_ready
    .groupby(['Disease', 'Treatment'])
    .agg({
        'symptom_standardized': lambda x: ', '.join(sorted(set(x)))
    })
    .reset_index()
)

# Rename for clarity
df_grouped = df_grouped.rename(columns={
    'symptom_standardized': 'Symptoms'
})

# Preview
df_grouped


Disease  \
0     acute respiratory distress syndrome   
1                              asbestosis   
2                              asbestosis   
3                              asbestosis   
4                              asbestosis   
5                           aspergillosis   
6                           aspergillosis   
7                           aspergillosis   
8                                  asthma   
9                                  asthma   
10                                 asthma   
11                                 asthma   
12                                 asthma   
13                         bronchiectasis   
14                         bronchiectasis   
15                         bronchiectasis   
16                         bronchiectasis   
17                          bronchiolitis   
18                          bronchiolitis   
19                          bronchiolitis   
20                          bronchiolitis   
21                             bronchitis   
22                             bronchitis   
23                     chronic bronchitis   
24                     chronic bronchitis   
25                     chronic bronchitis   
26                     chronic bronchitis   
27                          chronic cough   
28  chronic obstructive pulmonary disease   
29  chronic obstructive pulmonary disease   
30  chronic obstructive pulmonary disease   
31  chronic obstructive pulmonary disease   
32  chronic obstructive pulmonary disease   
33  chronic obstructive pulmonary disease   
34                              influenza   
35                           mesothelioma   
36                           mesothelioma   
37                                    nan   
38                                    nan   
39                                    nan   
40                                    nan   
41                              pneumonia   
42                              pneumonia   
43                              pneumonia   
44                              pneumonia   
45                           pneumothorax   
46                 pulmonary hypertension   
47                 pulmonary hypertension   
48            respiratory syncytial virus   
49                            sleep apnea   
50                            sleep apnea   
51                           tuberculosis   
52                           tuberculosis   
53                           tuberculosis   

                                Treatment  \
0                        consult a doctor   
1                        consult a doctor   
2                                     nan   
3                                  oxygen   
4              stay away from cold places   
5                            itraconazole   
6                                     nan   
7                                 surgery   
8                                 inhaler   
9                            itraconazole   
10                            mepolizumab   
11                                    nan   
12                             omalizumab   
13                            antibiotics   
14                      hypertonic saline   
15  intrapulmonary percussive ventilation   
16                                    nan   
17                       consult a doctor   
18                                inhaler   
19                                    nan   
20                      saline nose drops   
21                            antibiotics   
22                                    nan   
23                                inhaler   
24                                 oxygen   
25        steroids to reduce inflammation   
26                                  x-ray   
27                                    nan   
28                       consult a doctor   
29                                inhaler   
30                                    nan   
31                                 oxygen   
32               pulmonary rehabilitation   
33                          

In [26]:
df_grouped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Disease    54 non-null     object
 1   Treatment  54 non-null     object
 2   Symptoms   54 non-null     object
dtypes: object(3)
memory usage: 1.4+ KB


In [29]:
df_grouped['Disease'].nunique()

19

In [30]:
df_grouped['Disease'].value_counts()

Disease
chronic obstructive pulmonary disease    6
asthma                                   5
asbestosis                               4
pneumonia                                4
bronchiectasis                           4
bronchiolitis                            4
nan                                      4
chronic bronchitis                       4
tuberculosis                             3
aspergillosis                            3
mesothelioma                             2
bronchitis                               2
pulmonary hypertension                   2
sleep apnea                              2
chronic cough                            1
influenza                                1
pneumothorax                             1
respiratory syncytial virus              1
acute respiratory distress syndrome      1
Name: count, dtype: int64

### don’t want "nan" sitting as a disease category. Let’s explicitly drop those rows:

In [31]:
# ✅ Drop rows where Disease is NaN or literal 'nan'
df_grouped = df_grouped[df_grouped['Disease'].notna()]          # handles actual NaN
df_grouped = df_grouped[df_grouped['Disease'].str.lower() != 'nan']  # handles string "nan"

# Reset index for cleanliness
df_grouped = df_grouped.reset_index(drop=True)

# Preview cleaned dataset
df_grouped['Disease'].value_counts()


Disease
chronic obstructive pulmonary disease    6
asthma                                   5
chronic bronchitis                       4
pneumonia                                4
bronchiectasis                           4
bronchiolitis                            4
asbestosis                               4
tuberculosis                             3
aspergillosis                            3
bronchitis                               2
mesothelioma                             2
pulmonary hypertension                   2
sleep apnea                              2
chronic cough                            1
influenza                                1
pneumothorax                             1
respiratory syncytial virus              1
acute respiratory distress syndrome      1
Name: count, dtype: int64

In [32]:
df_grouped

Disease  \
0     acute respiratory distress syndrome   
1                              asbestosis   
2                              asbestosis   
3                              asbestosis   
4                              asbestosis   
5                           aspergillosis   
6                           aspergillosis   
7                           aspergillosis   
8                                  asthma   
9                                  asthma   
10                                 asthma   
11                                 asthma   
12                                 asthma   
13                         bronchiectasis   
14                         bronchiectasis   
15                         bronchiectasis   
16                         bronchiectasis   
17                          bronchiolitis   
18                          bronchiolitis   
19                          bronchiolitis   
20                          bronchiolitis   
21                             bronchitis   
22                             bronchitis   
23                     chronic bronchitis   
24                     chronic bronchitis   
25                     chronic bronchitis   
26                     chronic bronchitis   
27                          chronic cough   
28  chronic obstructive pulmonary disease   
29  chronic obstructive pulmonary disease   
30  chronic obstructive pulmonary disease   
31  chronic obstructive pulmonary disease   
32  chronic obstructive pulmonary disease   
33  chronic obstructive pulmonary disease   
34                              influenza   
35                           mesothelioma   
36                           mesothelioma   
37                              pneumonia   
38                              pneumonia   
39                              pneumonia   
40                              pneumonia   
41                           pneumothorax   
42                 pulmonary hypertension   
43                 pulmonary hypertension   
44            respiratory syncytial virus   
45                            sleep apnea   
46                            sleep apnea   
47                           tuberculosis   
48                           tuberculosis   
49                           tuberculosis   

                                Treatment  \
0                        consult a doctor   
1                        consult a doctor   
2                                     nan   
3                                  oxygen   
4              stay away from cold places   
5                            itraconazole   
6                                     nan   
7                                 surgery   
8                                 inhaler   
9                            itraconazole   
10                            mepolizumab   
11                                    nan   
12                             omalizumab   
13                            antibiotics   
14                      hypertonic saline   
15  intrapulmonary percussive ventilation   
16                                    nan   
17                       consult a doctor   
18                                inhaler   
19                                    nan   
20                      saline nose drops   
21                            antibiotics   
22                                    nan   
23                                inhaler   
24                                 oxygen   
25        steroids to reduce inflammation   
26                                  x-ray   
27                                    nan   
28                       consult a doctor   
29                                inhaler   
30                                    nan   
31                                 oxygen   
32               pulmonary rehabilitation   
33                                surgery   
34                            oseltamivir   
35                           chemotherapy   
36                       consult a doctor   
37                          